Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -0,0 +1,143 @@
+{-# LANGUAGE CPP #-}
+
+module Math.NumberTheory.Primes.Factorisation.QuadraticSieve
+  ( quadraticSieve
+  , findPairs
+  ) where
+
+#if __GLASGOW_HASKELL__ < 803
+import Data.Semigroup
+#endif
+import qualified Data.Vector as V
+import qualified Data.Vector.Mutable as MV
+import qualified Data.IntMap as I
+import qualified Data.IntSet as S
+import qualified Math.NumberTheory.Primes.IntSet as PS
+import Control.Monad
+import Control.Monad.ST
+import Data.Maybe
+import Data.Bifunctor
+import Math.NumberTheory.Roots
+import Math.NumberTheory.Primes
+import Math.NumberTheory.Moduli.Sqrt
+import Math.NumberTheory.Utils.FromIntegral
+
+data SignedPrimeIntSet = SignedPrimeIntSet
+  { sign :: !Bool
+  , primeSet :: !PS.PrimeIntSet
+  } deriving (Show)
+
+data BoolOrPrime = Bool Bool | PrimeInt (Prime Int)
+
+insert :: Prime Int -> SignedPrimeIntSet -> SignedPrimeIntSet
+insert prime (SignedPrimeIntSet s ps) = SignedPrimeIntSet s (prime `PS.insert` ps)
+
+nonZero :: SignedPrimeIntSet -> Maybe BoolOrPrime
+nonZero (SignedPrimeIntSet s ps) = case PS.minView ps of
+  Just (prime, _)  -> Just (PrimeInt prime)
+  Nothing          -> if s then Just (Bool True) else Nothing
+
+member :: BoolOrPrime -> SignedPrimeIntSet -> Bool
+member value (SignedPrimeIntSet s ps) = case value of
+  Bool b     -> s == b
+  PrimeInt p -> p `PS.member` ps
+
+xor :: SignedPrimeIntSet -> SignedPrimeIntSet -> SignedPrimeIntSet
+xor (SignedPrimeIntSet s1 ps1) (SignedPrimeIntSet s2 ps2) = SignedPrimeIntSet (s1 /= s2) ((ps1 PS.\\ PS.unPrimeIntSet ps2) <> (ps2 PS.\\ PS.unPrimeIntSet ps1))
+
+-- Given an odd positive composite Integer n and Int parametres b and t,
+-- the Quadratic Sieve attempt to decompose n into smaller factors p and q.
+quadraticSieve :: Integer -> Int -> Int -> Integer
+quadraticSieve n b t = findFactor n $ findPairs n b t
+
+findFactor :: Integer -> [(Integer, Integer)] -> Integer
+findFactor _ [] = error "Parametres are not large enough."
+findFactor n ((x, y) : otherPairs) = if factor == 1 || factor == n then findFactor n otherPairs else factor
+  where
+    factor = gcd (x - y) n
+
+findPairs :: Integer -> Int -> Int -> [(Integer, Integer)]
+findPairs n b t = runST $ do
+  let
+    factorBase = [nextPrime 2..precPrime b]
+    squareRoot = integerSquareRoot n
+    sievingFunction j = j * j - n
+    startingPoint = squareRoot - intToInteger t `div` 2
+    sievingInterval = generateInterval sievingFunction startingPoint t
+  sievingIntervalM <- V.thaw sievingInterval
+  -- Perform sieving
+  smoothSieveM sievingIntervalM factorBase n startingPoint
+  sievingIntervalF <- V.unsafeFreeze sievingIntervalM
+  -- Filter smooth numbers
+  let
+    indexedFactorisations = V.toList (findSmoothNumbers sievingIntervalF)
+    solutionBasis = gaussianElimination indexedFactorisations
```

Comment:
I tried the optimization, which removes rows with primes, which does not appear in any other row. It seems to work quite well, actually. The example in the executable computes 32714 rows of `indexedFactorisations` and takes 842 seconds in total. But after removing rows with primes, which appear only once, we are left with only 13634 rows and factorisation completes in 270 seconds. 

Here is a draft implementation:

```haskell
...
    indexedFactorisations' = V.toList (findSmoothNumbers sievingIntervalF)
    onlyOnce = appearsOnlyOnce $ map (primeSet . snd) indexedFactorisations'
    indexedFactorisations = filter (\(_, SignedPrimeIntSet _ xs) -> PS.disjoint xs onlyOnce) indexedFactorisations'
...

-- | Find all primes, which appear only once in the input list.
appearsOnlyOnce :: [PS.PrimeIntSet] -> PS.PrimeIntSet
appearsOnlyOnce = fst . foldl' go (mempty, mempty)
  where
    go (onlyOnce, atLeastOnce) x =
      ((onlyOnce PS.\\ PS.unPrimeIntSet x) <> (x PS.\\ PS.unPrimeIntSet atLeastOnce), atLeastOnce <> x)
```

Reply:
I must have done something wrong before. This seems to be faster